### PR TITLE
[PyTorch] Fix get_swa_mask() for padding masks

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR fixes the mask generation for sliding window in `UnfusedDotProductAttention`. It fixes the logic for padding and arbitrary masks in `get_swa_mask()`, adds more docstring, refactors the call site, and adds more testing in the unit tests.

Fixes #1271 

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Improve the logic in `get_swa_mask()` and its call site

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by Marks101** [comment]
Hi @cyanguwa,
great, I like the idea to have all the masking logic at one place 👍 
I just tested this and found a problem with cross attention:
```python
        if "padding" in attn_mask_type:
            if max_seqlen_q == max_seqlen_kv:
                attention_mask = torch.logical_or(
>                   attention_mask.squeeze(1).unsqueeze(3), attention_mask
                )
E               AttributeError: 'tuple' object has no attribute 'squeeze'
```
The code in `UnfusedDotProductAttention` made these lines dependent on the `attention_type`.

**Comment by cyanguwa** [comment]
> Hi @cyanguwa, great, I like the idea to have all the masking logic at one place 👍 I just tested this and found a problem with cross attention:
> 
> ```python
>         if "padding" in attn_mask_type:
>             if max_seqlen_q == max_seqlen_kv:
>                 attention_mask = torch.logical_or(
> >                   attention_mask.squeeze(1).unsqueeze(3), attention_mask
>                 )
> E               AttributeError: 'tuple' object has no attribute 'squeeze'
> ```
> 
> The code in `UnfusedDotProductAttention` made these lines dependent on the `attention_type`.

Yes, I think I should use `if attention_type == "self"` here because there could be cross-attention cases where `max_seqlen_q == max_seqlen_kv` and `actual_seqlen_q != actual_seqlen_kv`. I'll go through `attention.py` and see if there're other places I should use `attention_type` instead.

Let me know if you observe any other issues too! :) Thanks!

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -1030,9 +1030,24 @@ def get_swa_mask(
...
-    For "`causal`" mask type, the sliding window diagonal is aligned to the top left corner,
-    and for other mask types, the bottom right corner.
+    Convert sliding window `window_size` to an equivalent "`arbitrary`" mask. Requirements for the
+    shapes of `attention_mask` given an `attn_mask_type` are the same as in DotProductAttention.
+    For "`causal`" and "`padding_causal`" mask types, the sliding window diagonal is aligned to the
+    top left corner of the softmax matrix; for others, the bottom right corner. Note that when padding
+    is applied, the bottom right corner comes from the [actual_seqlen_q[i], actual_seqlen_kv[i]] matrix,
+    for each batch i, not the [max_seqlen_q, max_seqlen_kv] matrix.::
+
+       attn_mask_type              output shape                                 diagonal alignment
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:1040]: Comment

**Review by xrennvidia — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**